File: C++/HashSet.cpp
struct Node{
    int val;
    Node *next;
    Node(int val){
        this->val=val;
    }
};
class HashSet{
private:
    Node **key;
    int size;
    HashSet(int size){
        this->size=size;
        key=new Node*[size];
        for(int i=0;i<size;i++){
            key[i]=0;
        }       
    }
    ~HashSet(){
        Node *next;
        for(int i=0;i<size;i++){
            if(key[i]!=0){
                Node *curr=key[i];
                while(curr!=0){     
                    next=curr.next;
                    delete(curr);
                    curr=next;
                }
                key[i]=0;
            }
        }
    }
    int hashCode(int value){
        return value%size;
    }
public:
    bool add(int value){
        int k=hashCode(value);
        Node *curr=key[k];
        while(curr!=0){
            if(curr->val==value){
                return true;
            }
            curr=curr->nextl
        }
        Node *head=new Node(value);
        head->next=key[k];
        key[k]=head;
        return false;       
    }
    bool remove(int value){
        int k=hashCode(value);
        Node *curr=key[k];
        if(curr->val==value){
            key[k]=key[k]->next;
            delete(curr);
            return true;
        }
        while(curr->next->val!=value){
            curr=curr->val;
        }
        Node *temp=curr->next;
        curr->next=temp->next;
        delete(temp);
        return true;
    }
};

File: C++/HashSet.java
class MapNode<T>{
	T key;
	MapNode<T> next;
	MapNode(T node){
		key = node;
		next = null;
	}
}
public class HashSet<T>{
	int size;
	MapNode<T> map[];
	HashSet(int size){
		this.size = size;
		map = new MapNode[size];
		for(int i=0;i<size;i++)
			map[i]=null;
	}
	HashSet(){
		this.size = 1000;
		map = new MapNode[size];
		for(int i=0;i<size;i++)
			map[i]=null;		
	}
	boolean put(T node){
		int key = (node.hashCode()%size + size)%size;				
			MapNode<T> temp = map[key];
			while(temp!=null){;
				if(temp.key.equals(node)){
					return false;
				}
				temp = temp.next;
			}
			temp = map[key];
			map[key] = new MapNode<T>(node);
			map[key].next=temp;
			return true;
	}
	T get(T node){
		int key = (node.hashCode()%size + size)%size;
		MapNode<T> temp = map[key];
		while(temp!=null){
			if(temp.key.equals(node)){
				return temp.key;
			}
			temp = temp.next;
		}
		return null;		
	}
	boolean remove(T node){
		int key = (node.hashCode()%size + size)%size;
		MapNode<T> temp = map[key];
		MapNode<T> pre = null;
		while(temp!=null){
			if(temp.key.equals(node)){
				break;
			}
			pre = temp;
			temp = temp.next;
		}
		if(temp!=null){
			if(pre==null){
				map[key]=map[key].next;
			}else{
				pre.next=pre.next.next;
			}
			return true;
		}
		return false;
	}
}